### PR TITLE
Encode special characters for generated shell command

### PR DESCRIPTION
**Pull request by BradyShober**
See [JENKINS-35278](https://issues.jenkins-ci.org/browse/JENKINS-35278).

### Proposed changelog entries

* Entry 1: Issue, Use full URL character encoding for the generated JNLP agent launch string

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by BradyShober** [comment]
@jenkinsci/core-pr-reviewers

**Review by res0nance — APPROVED** [review]

File: core/src/main/java/hudson/Util.java
Comment (timja) [inline core/src/main/java/hudson/Util.java:924]: This is in a util function not sure why jnlp is mentioned here? It can be used anywhere

Comment (BradyShober) [reply core/src/main/java/hudson/Util.java:924]: I was mainly trying to differentiate this from the existing function, I'll update the wording on the comment

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by daniel-beck** [comment]
Does the existing script work for Windows batch, and if so, does the new one too? `%` chars are metacharacters there.

File: core/src/main/java/hudson/Util.java
Comment (daniel-beck) [inline core/src/main/java/hudson/Util.java:929]:
```suggestion
     * @since TODO
     */
```

(Unless we prefer `@Restricted(NoExternalUse.class)`)

**Comment by BradyShober** [comment]
> Does the existing script work for Windows batch, and if so, does the new one too? `%` chars are metacharacters there.

@daniel-beck Yes, the existing version worked for Windows at least with the case described in the issue with parentheses in the agent name. The new version also works on Windows with the % chars

File: core/src/main/java/hudson/Util.java
Comment (Wadeck) [inline core/src/main/java/hudson/Util.java:898]:
```suggestion
        String raw = "               0123456789       ABCDEFGHIJKLMNOPQRSTUVWXYZ      abcdefghijklmnopqrstuvwxyz";
        //            !"#$%&'()*+,-./0123456789:;<=>?@                          [\]^_`                          {|}~
        //  ^--so these are encoded
```

The idea with the comments in the static section for uriMap was to start from the ASCII table as a list and to display which ones will be kept and which ones will be encoded. With your proposal we lost this information as the characters are in an arbitrary order (and could be confusing).

File: core/src/main/java/hudson/Util.java
Comment (Wadeck) [inline core/src/main/java/hudson/Util.java:926]:
```suggestion
     * All characters outside numbers and letters without diacritic are encoded.
```

Otherwise you need to add "control characters" (0-32) and all characters after code 122 (z).

**Review by Wadeck — COMMENTED** [review]
Thank you for the contribution @BradyShober.

I proposed two changes to improve the comments around your code. The code itself seems to be fine :)

In addition, I tested with Bash and Batch if everything is working as expected. 
Key points:
- If I have a job called `regu+lar+`, it will be encoded `regu%2Blar%2B`. If I have an Env variable called `2Blar` my copy-pasted command will fail. Before this PR, that job name will not have been changed (and thus, not having troubles on Windows)
- Before this PR, a job with spaces would result in same issue with the encoded space instead of plus, leading to command not being executable.
- In order to be impacted, you need to have env variables with the 2 first letters being hexadecimal as hard constraint, which is not the case "by default".
- In addition, if the env variable is not present in Batch, there will be no replacement and thus it will work fine.
- You are improving the situation for Unix user with job names containing parenthesis, which are fairly more frequent than the Env var starting with hex under Windows.

So conclusion, you are overall improving the situation for most common cases 🎉 

Side note: thanks for the test!

_(I will approve as soon as you applied my proposals, especially the ascii list comment)_

**Review by oleg-nenashev — APPROVED** [review]
LGTM, thank you @BradyShober !

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Review by Wadeck — APPROVED** [review]
🎉

**Pull request merged**
